### PR TITLE
sources: make submodule recursion configurable

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `./runtests.sh static`?
- [X] Have you successfully run `./runtests.sh tests/unit`?

-----
https://bugs.launchpad.net/snapcraft/+bug/1957767

Add new parameter `source_recurse_submodules`

Example snapcraft.yaml:
```
parts:
    git:
        source: https://github.com/snapcore/snapcraft
        source-type: git
        source-commit: "HEAD"
        source-recurse-submodules: False
```

### CODE REVIEW

File: snapcraft/internal/sources/__init__.py [hunk 1]
```diff
@@ -65,6 +65,10 @@
     When building, Snapcraft will set the working directory to be this
     subdirectory within the source.
 
+  - source-recurse-submodule: <boolean>
+
+    Snapcraft will recursively fetch submodules from the source tree by default.
```
Comment (cmatsuoka) [inline snapcraft/internal/sources/__init__.py:70]: Perhaps we could rephrase this a little bit to tell what the option does, and then tell what the default is. Something like "Control whether submodules will be recursively fetched or not. The default is to recursively fetch submodules from the source tree."

File: tests/unit/sources/test_git.py [hunk 2]
```diff
@@ -270,6 +270,28 @@ def test_pull_commit(self):
...
+            "git://my-source", "source_dir", source_recurse_submodules=True
+        )
+
+        git.pull()
+
+        self.mock_run.assert_called_once_with(
+            ["git", "clone", "--recursive", "git://my-source", "source_dir"]
+        )
+
+    def test_pull_with_recurse_submodules_false(self):
```
Comment (cmatsuoka) [inline tests/unit/sources/test_git.py:284]: We should also test the default value to make sure it's enabling recursion.

**Review by cmatsuoka — CHANGES_REQUESTED** [review]
The implementation looks good, I just added some minor notes about the help text and testing the default value.

About the commit message format: could you amend the commit to mention the subsystem being worked on in the subject line (something like `sources: make submodule recursion configurable`) and add some explanatory text? You can also add `LP: #nnnnn` with the number of the Launchpad bug to have it updated automatically when this PR is merged.  Example: see commit 83d54e5

**Comment by mr-cal** [comment]
Thanks, Claudio.  I updated based on your feedback.

**Comment by cmatsuoka** [comment]
> Should I also update `schema/snapcraft.json`?

Yes, the schema file must be updated to be consistent with the available options (otherwise schema validation will fail when using the new property).

**Comment by mr-cal** [comment]
Updated schema.

**Review by cmatsuoka — CHANGES_REQUESTED** [review]
Thanks, the changes look good to me. We don't usually force-push after the review starts but in this case it was necessary to amend the commit subject line. Could you do it once more to add a short description message to the commit with some more information about the change (e.g. "Git clone recursion can now be configured using the xyz option. If not declared, the previous behavior will be kept" or something along those lines. This helps us to understand what the commit was about if we need to revert or cherry-pick it later.

**Comment by cmatsuoka** [comment]
Also I don't think it's a strict requirement in this project but I usually add a "Signed-off-by" line to my commits (using `git commit -s`).

**Comment by sergiusens** [comment]
Looks pretty good.

I am having second thoughts about this, would you think that in the feature this could be expanded to allow for only getting specific submodules?

Mind rebasing to latest changes, and maybe amending "fix: " to something more functional like "static: "?

**Comment by cmatsuoka** [comment]
I'm just wondering if the new proposal is practical to implement, does git provide per-module recursion control, or just the catch-all `--recurse-submodules`?

**Comment by mr-cal** [comment]
Git does provide per-module recursion control.  

For clarity, I put the new proposal under a new pull request, https://github.com/snapcore/snapcraft/pull/3629.  I hope it will satisfies any practical use case.

**Comment by cmatsuoka** [comment]
Superseded by PR #3629

**Pull request closed without merging**